fri/src/prover/mod.rs

```diff
@@ -230,47 +250,44 @@ where
-        let folding_factor = self.options.folding_factor();
 
-        // for all FRI layers, except the last one, record tree root, determine a set of query
-        // positions, and query the layer at these positions.
         let mut layers = Vec::with_capacity(self.layers.len());
-        for i in 0..self.layers.len() - 1 {
-            positions = fold_positions(&positions, domain_size, folding_factor);
-
-            // sort of a static dispatch for folding_factor parameter
-            let proof_layer = match folding_factor {
-                2 => query_layer::<B, E, H, 2>(&self.layers[i], &positions),
-                4 => query_layer::<B, E, H, 4>(&self.layers[i], &positions),
-                8 => query_layer::<B, E, H, 8>(&self.layers[i], &positions),
-                16 => query_layer::<B, E, H, 16>(&self.layers[i], &positions),
-                _ => unimplemented!("folding factor {} is not supported", folding_factor),
-            };
-
-            layers.push(proof_layer);
-            domain_size /= folding_factor;
-        }
 
-        // use the remaining polynomial values directly as proof; last layer values contain
-        // remainder in transposed form - so, we un-transpose it first
-        let last_values = &self.layers[self.layers.len() - 1].evaluations;
-        let mut remainder = E::zeroed_vector(last_values.len());
-        let n = last_values.len() / folding_factor;
-        for i in 0..n {
-            for j in 0..folding_factor {
-                remainder[i + n * j] = last_values[i * folding_factor + j];
+        if !self.layers.is_empty() {
+            let mut positions = positions.to_vec();
+            let mut domain_size = self.layers[0].evaluations.len();
+            let folding_factor = self.options.folding_factor();
+
+            // for all FRI layers, except the last one, record tree root, determine a set of query
+            // positions, and query the layer at these positions.
+            for i in 0..self.layers.len() {
+                positions = fold_positions(&positions, domain_size, folding_factor);
+
+                // sort of a static dispatch for folding_factor parameter
+                let proof_layer = match folding_factor {
+                    2 => query_layer::<B, E, H, 2>(&self.layers[i], &positions),
+                    4 => query_layer::<B, E, H, 4>(&self.layers[i], &positions),
+                    8 => query_layer::<B, E, H, 8>(&self.layers[i], &positions),
+                    16 => query_layer::<B, E, H, 16>(&self.layers[i], &positions),
+                    _ => unimplemented!("folding factor {} is not supported", folding_factor),
+                };
+
+                layers.push(proof_layer);
+                domain_size /= folding_factor;
             }
         }
 
+        // use the remaining polynomial values directly as proof; last layer values contain
+        // remainder.
+        let last_values = self.remainder.0.clone();
```

Comment:
nit: let's change `last_values` to `remainder`?

Reply:
Yes, that's more consistent.